builtin/game/player_on_step.lua

```diff
@@ -1,9 +1,57 @@
 
-core.register_globalstep(function(...)
+-- stuff for core.register_player_on_slow_step
+local time = 0.0
+local time_next = math.huge
+local step_sizes = {}
+local expiring_times = {}
+
+-- override core.register_player_on_slow_step to also store the step size in
+-- step_sizes and update expiring_times and time_next
+local old_register_player_on_slow_step = core.register_player_on_slow_step
+
+function core.register_player_on_slow_step(step_size, func)
+	old_register_player_on_slow_step(func)
+
+	local index = #core.registered_player_on_slow_steps
+	step_sizes[index] = step_size
+	expiring_times[index] = step_size
+	time_next = math.min(time_next, step_size)
+end
+
+core.register_globalstep(function(dtime)
 	local players = core.get_connected_players()
+
+	-- core.register_player_on_step
 	for i = 1, #players do
-		-- callback mode 5: RUN_CALLBACKS_MODE_OR_SC
-		-- if any callback returns true, the player is kicked or similar
-		core.run_callbacks(core.registered_player_on_steps, 5, players[i], ...)
+		-- the callback mode does not matter
+		core.run_callbacks(core.registered_player_on_steps, 0, players[i], dtime)
+	end
+
+	-- core.register_player_on_slow_step
+	time = time + dtime
+
+	if time < time_next then
+		return
+	end
+
+	time_next = math.huge
+
+	for i = 1, #expiring_times do
```

Comment:
why not reuse minetest.after for all this?

Reply:
Actually, this works pretty much like `minetest.after`.
But `minetest.after` can't be used directly. We need to pass the current player list, so we have to call the callback functions directly.
It would be possible to have a .after function call itself again after `time_next`, but then we would have to call `get_connected_players` twice int that globalstep step.